File: lib/features/activity/presentation/activity_splits_card.dart
import 'package:flutter/material.dart';

import '../../../core/presentation/unit_formatter.dart';
import '../../../domain/insights/activity_splits.dart';
import '../../../domain/preferences/unit_system.dart';
import '../../../l10n/app_localizations.dart';
import '../../../ui/components/ov_card.dart';
import '../../../ui/theme/app_colors.dart';
import 'activity_split_distance_label.dart';

/// The splits card on the activity detail screen: one row per split (or per
/// device lap), with a bar whose length tracks the split's pace so a slow
/// kilometre is visible without reading a single number.
///
/// The header states the PROVENANCE, and the estimated case says out loud that
/// the identical pace on every row is an artefact of missing data — the whole
/// point of keeping [SplitSource] on the result.
class ActivitySplitsCard extends StatelessWidget {
  const ActivitySplitsCard({
    super.key,
    required this.splits,
    required this.formatter,
    required this.splitDistanceMeters,
    required this.slowestPaceSeconds,
    required this.fastestPaceSeconds,
  });

  final ActivitySplits splits;
  final UnitFormatter formatter;
  final double splitDistanceMeters;

  /// The activity's slowest and fastest split, in seconds per kilometre — the
  /// bar scale, folded at load time by the view-model. A ratio between splits,
  /// so the unit it is expressed in cancels out; the numbers the row PRINTS are
  /// still formatted in the user's own units.
  final double? slowestPaceSeconds;
  final double? fastestPaceSeconds;

  @override
  Widget build(BuildContext context) {
    final l10n = AppLocalizations.of(context);
    final theme = Theme.of(context);
    final rows = splits.splits;
    if (rows.isEmpty) return const SizedBox.shrink();

    final unitMeters = switch (formatter.unitSystem()) {
      UnitSystem.metric => 1000.0,
      UnitSystem.imperial => 1609.344,
    };

    final (title, subtitle) = switch (splits.source) {
      SplitSource.deviceLaps => (
          l10n.activitySplitsLapsTitle,
          l10n.activitySplitsLapsBody,
        ),
      SplitSource.route || SplitSource.speedSamples => (
          l10n.activitySplitsDerivedTitle(
            splitDistanceLabel(l10n, formatter, splitDistanceMeters),
          ),
          null,
        ),
      SplitSource.estimated => (
          l10n.activitySplitsEstimatedTitle,
          l10n.activitySplitsEstimatedBody,
        ),
    };
    final columnLabel = splits.source == SplitSource.deviceLaps
        ? l10n.activitySplitsHeaderLap
        : l10n.activitySplitsHeaderSplit;

    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              children: [
                Icon(
                  Icons.timeline_outlined,
                  color: theme.colorScheme.onSurfaceVariant,
                  size: 20,
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: Text(title, style: theme.textTheme.titleMedium),
                ),
              ],
            ),
            if (subtitle != null) ...[
              const SizedBox(height: 4),
              Text(
                subtitle,
                style: theme.textTheme.bodySmall
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              ),
            ],
            const SizedBox(height: 12),
            for (final split in rows) ...[
              _SplitRow(
                split: split,
                columnLabel: columnLabel,
                formatter: formatter,
                unitMeters: unitMeters,
                slowestPaceSeconds: slowestPaceSeconds,
                fastestPaceSeconds: fastestPaceSeconds,
                // The estimated source gives every split the same pace: a bar
                // chart of it would be a straight line masquerading as a
                // measurement. Show the numbers, drop the bar.
                showBar: splits.source != SplitSource.estimated,
              ),
              const SizedBox(height: 10),
            ],
          ],
        ),
      ),
    );
  }
}

class _SplitRow extends StatelessWidget {
  const _SplitRow({
    required this.split,
    required this.columnLabel,
    required this.formatter,
    required this.unitMeters,
    required this.slowestPaceSeconds,
    required this.fastestPaceSeconds,
    required this.showBar,
  });

  final ActivitySplit split;
  final String columnLabel;
  final UnitFormatter formatter;
  final double unitMeters;
  final double? slowestPaceSeconds;
  final double? fastestPaceSeconds;
  final bool showBar;

  @override
  Widget build(BuildContext context) {
    final l10n = AppLocalizations.of(context);
    final theme = Theme.of(context);
    final muted = theme.textTheme.bodySmall
        ?.copyWith(color: theme.colorScheme.onSurfaceVariant);

    final pace = formatter.averagePace(
      split.distanceMeters,
      split.elapsed.inMilliseconds,
    );
    final delta = split.paceDeltaSecondsPerUnit(unitMeters);

    return Semantics(
      container: true,
      label: _semanticsLabel(l10n),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.baseline,
            textBaseline: TextBaseline.alphabetic,
            children: [
              SizedBox(
                width: 24,
                child: Text(
                  '${split.index}',
                  style: theme.textTheme.titleSmall,
                ),
              ),
              Expanded(
                child: Text(
                  _distanceText(l10n),
                  style: theme.textTheme.bodyMedium,
                ),
              ),
              Text(
                formatter.duration(split.elapsed.inMilliseconds),
                style: muted,
              ),
              const SizedBox(width: 12),
              SizedBox(
                width: 92,
                child: Text(
                  pace?.text ?? '--',
                  textAlign: TextAlign.end,
                  style: theme.textTheme.bodyMedium
                      ?.copyWith(fontWeight: FontWeight.w600),
                ),
              ),
            ],
          ),
          if (showBar) ...[
            const SizedBox(height: 4),
            Padding(
              padding: const EdgeInsets.only(left: 24),
              child: _PaceBar(
                // Per kilometre, like the scale it is measured against.
                paceSeconds: split.paceSecondsPerUnit(1000.0),
                slowestPaceSeconds: slowestPaceSeconds,
                fastestPaceSeconds: fastestPaceSeconds,
                fasterThanAverage: delta != null && delta < 0,
              ),
            ),
          ],
          const SizedBox(height: 4),
          Padding(
            padding: const EdgeInsets.only(left: 24),
            child: Text(_detailLine(delta), style: muted),
          ),
        ],
      ),
    );
  }

  String _distanceText(AppLocalizations l10n) {
    final distance = formatter.distance(split.distanceMeters).text;
    // A partial split is short ON PURPOSE — say so, or it reads as a bad fix.
    return split.isPartial ? '$distance (${l10n.activitySplitsPartial})' : distance;
  }

  /// avg HR · elevation ± · pace delta. Each piece is dropped when its datum is
  /// missing rather than rendered as a zero: a treadmill split has NO elevation,
  /// which is not the same claim as "flat".
  String _detailLine(double? delta) {
    final parts = <String>[];
    final bpm = split.averageHeartRateBpm;
    if (bpm != null) parts.add(formatter.heartRate(bpm).text);

    final gain = split.elevationGainMeters;
    final loss = split.elevationLossMeters;
    if (gain != null && loss != null) {
      parts.add(
        '↑ ${formatter.elevation(gain).text}  ↓ ${formatter.elevation(loss).text}',
      );
    }
    if (delta != null) parts.add(_formatDelta(delta));
    return parts.join('  ·  ');
  }

  String _semanticsLabel(AppLocalizations l10n) {
    final pace = formatter.averagePace(
      split.distanceMeters,
      split.elapsed.inMilliseconds,
    );
    final delta = split.paceDeltaSecondsPerUnit(unitMeters);
    final buffer = StringBuffer(
      '$columnLabel ${split.index}, '
      '${formatter.distance(split.distanceMeters).text}, '
      '${formatter.duration(split.elapsed.inMilliseconds)}',
    );
    if (pace != null) buffer.write(', ${pace.text}');
    if (delta != null && delta.round() != 0) {
      final magnitude = _formatSeconds(delta.abs());
      buffer.write(
        ', ${delta < 0 ? l10n.activitySplitsFaster(magnitude) : l10n.activitySplitsSlower(magnitude)}',
      );
    }
    return buffer.toString();
  }
}

/// `-0:08` / `+0:12` — signed minutes:seconds against the activity average.
String _formatDelta(double deltaSeconds) {
  final rounded = deltaSeconds.round();
  if (rounded == 0) return '±0:00';
  final sign = rounded < 0 ? '−' : '+';
  return '$sign${_formatSeconds(rounded.abs().toDouble())}';
}

String _formatSeconds(double seconds) {
  final total = seconds.round();
  final minutes = total ~/ 60;
  final rest = total % 60;
  return '$minutes:${rest.toString().padLeft(2, '0')}';
}

/// A pace bar: the slowest split in the activity fills the track, the fastest
/// leaves it visibly shorter. Deliberately NOT zero-based — the interesting
/// range of a run is the few percent between its fastest and slowest kilometre,
/// and a zero-based bar squashes that into invisibility. The floor is 25% of
/// the track so the fastest split still reads as a bar, not as nothing.
class _PaceBar extends StatelessWidget {
  const _PaceBar({
    required this.paceSeconds,
    required this.slowestPaceSeconds,
    required this.fastestPaceSeconds,
    required this.fasterThanAverage,
  });

  final double? paceSeconds;
  final double? slowestPaceSeconds;
  final double? fastestPaceSeconds;
  final bool fasterThanAverage;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final pace = paceSeconds;
    final slowest = slowestPaceSeconds;
    final fastest = fastestPaceSeconds;
    if (pace == null || slowest == null || fastest == null) {
      return const SizedBox(height: 6);
    }

    const minFraction = 0.25;
    final span = slowest - fastest;
    final fraction = span <= 0
        ? 1.0
        : minFraction + (1 - minFraction) * ((pace - fastest) / span);

    return ClipRRect(
      borderRadius: BorderRadius.circular(3),
      child: LinearProgressIndicator(
        value: fraction.clamp(minFraction, 1.0),
        minHeight: 6,
        backgroundColor: theme.colorScheme.surfaceContainerHighest,
        valueColor: AlwaysStoppedAnimation<Color>(
          fasterThanAverage
              ? AppColors.workout
              : theme.colorScheme.primary.withValues(alpha: 0.55),
        ),
      ),
    );
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/features/activity/ActivitySplitsCard.kt
package tech.mmarca.openvitals.features.activity

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Timeline
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlin.math.roundToLong
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import tech.mmarca.openvitals.domain.insights.ActivitySplit
import tech.mmarca.openvitals.domain.insights.ActivitySplits
import tech.mmarca.openvitals.domain.insights.SplitSource
import tech.mmarca.openvitals.domain.preferences.UnitSystem
import tech.mmarca.openvitals.ui.components.OpenVitalsCard
import tech.mmarca.openvitals.ui.theme.WorkoutColor

/**
 * The splits card on the activity detail screen: one row per split (or per
 * device lap), with a bar whose length tracks the split's pace so a slow
 * kilometre is visible without reading a single number.
 *
 * The header states the PROVENANCE, and the estimated case says out loud that
 * the identical pace on every row is an artefact of missing data — the whole
 * point of keeping [SplitSource] on the result.
 */
@Composable
internal fun ActivitySplitsCard(
    splits: ActivitySplits,
    splitDistanceMeters: Double,
    slowestPaceSeconds: Double?,
    fastestPaceSeconds: Double?,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
) {
    val rows = splits.splits
    if (rows.isEmpty()) return

    val unitMeters = when (unitFormatter.unitSystem()) {
        UnitSystem.METRIC -> 1000.0
        UnitSystem.IMPERIAL -> 1609.344
    }

    val title: String
    val subtitle: String?
    when (splits.source) {
        SplitSource.DEVICE_LAPS -> {
            title = stringResource(R.string.activity_splits_laps_title)
            subtitle = stringResource(R.string.activity_splits_laps_body)
        }
        SplitSource.ROUTE, SplitSource.SPEED_SAMPLES -> {
            title = stringResource(
                R.string.activity_splits_derived_title,
                splitDistanceLabel(unitFormatter, splitDistanceMeters),
            )
            subtitle = null
        }
        SplitSource.ESTIMATED -> {
            title = stringResource(R.string.activity_splits_estimated_title)
            subtitle = stringResource(R.string.activity_splits_estimated_body)
        }
    }

    OpenVitalsCard(modifier = modifier) {
        Column(modifier = Modifier.padding(16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Outlined.Timeline,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(20.dp),
                )
                Spacer(Modifier.width(12.dp))
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.weight(1f),
                )
            }
            if (subtitle != null) {
                Spacer(Modifier.height(4.dp))
                Text(
                    text = subtitle,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Spacer(Modifier.height(12.dp))
            rows.forEach { split ->
                SplitRow(
                    split = split,
                    unitFormatter = unitFormatter,
                    unitMeters = unitMeters,
                    slowestPaceSeconds = slowestPaceSeconds,
                    fastestPaceSeconds = fastestPaceSeconds,
                    // The estimated source gives every split the same pace: a
                    // bar chart of it would be a straight line masquerading as
                    // a measurement. Show the numbers, drop the bar.
                    showBar = splits.source != SplitSource.ESTIMATED,
                )
                Spacer(Modifier.height(10.dp))
            }
        }
    }
}

@Composable
private fun SplitRow(
    split: ActivitySplit,
    unitFormatter: UnitFormatter,
    unitMeters: Double,
    slowestPaceSeconds: Double?,
    fastestPaceSeconds: Double?,
    showBar: Boolean,
) {
    val muted = MaterialTheme.typography.bodySmall
    val mutedColor = MaterialTheme.colorScheme.onSurfaceVariant

    val pace = unitFormatter.averagePace(split.distanceMeters, split.elapsedMs)
    val delta = split.paceDeltaSecondsPerUnit(unitMeters)

    // A partial split is short ON PURPOSE — say so, or it reads as a bad fix.
    val distance = unitFormatter.distance(split.distanceMeters).text
    val distanceText = if (split.isPartial) {
        "$distance (${stringResource(R.string.activity_splits_partial)})"
    } else {
        distance
    }

    Column {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(modifier = Modifier.width(24.dp)) {
                Text(
                    text = split.index.toString(),
                    style = MaterialTheme.typography.titleSmall,
                )
            }
            Text(
                text = distanceText,
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier.weight(1f),
            )
            Text(
                text = unitFormatter.duration(split.elapsedMs),
                style = muted,
                color = mutedColor,
            )
            Spacer(Modifier.width(12.dp))
            Text(
                text = pace?.text ?: "--",
                textAlign = TextAlign.End,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.width(92.dp),
            )
        }
        if (showBar) {
            Spacer(Modifier.height(4.dp))
            PaceBar(
                // Per kilometre, like the scale it is measured against.
                paceSeconds = split.paceSecondsPerUnit(1000.0),
                slowestPaceSeconds = slowestPaceSeconds,
                fastestPaceSeconds = fastestPaceSeconds,
                fasterThanAverage = delta != null && delta < 0,
                modifier = Modifier.padding(start = 24.dp),
            )
        }
        Spacer(Modifier.height(4.dp))
        Text(
            text = splitDetailLine(split, delta, unitFormatter),
            style = muted,
            color = mutedColor,
            modifier = Modifier.padding(start = 24.dp),
        )
    }
}

/**
 * avg HR · elevation ± · pace delta. Each piece is dropped when its datum is
 * missing rather than rendered as a zero: a treadmill split has NO elevation,
 * which is not the same claim as "flat".
 */
@Composable
private fun splitDetailLine(
    split: ActivitySplit,
    delta: Double?,
    unitFormatter: UnitFormatter,
): String {
    val parts = mutableListOf<String>()
    split.averageHeartRateBpm?.let { bpm ->
        parts.add(unitFormatter.heartRate(bpm.roundToLong()).text)
    }
    val gain = split.elevationGainMeters
    val loss = split.elevationLossMeters
    if (gain != null && loss != null) {
        parts.add(
            "↑ ${unitFormatter.elevation(gain).text}  ↓ ${unitFormatter.elevation(loss).text}",
        )
    }
    if (delta != null && delta.roundToInt() != 0) {
        val magnitude = formatSplitDeltaSeconds(abs(delta))
        parts.add(
            if (delta < 0) {
                stringResource(R.string.activity_splits_faster, magnitude)
            } else {
                stringResource(R.string.activity_splits_slower, magnitude)
            },
        )
    }
    return parts.joinToString("  ·  ")
}

/** `0:08` / `1:12` — the delta's magnitude as minutes:seconds. */
private fun formatSplitDeltaSeconds(seconds: Double): String {
    val total = seconds.roundToInt()
    val minutes = total / 60
    val rest = total % 60
    return "$minutes:${rest.toString().padStart(2, '0')}"
}

/**
 * A pace bar: the slowest split in the activity fills the track, the fastest
 * leaves it visibly shorter. Deliberately NOT zero-based — the interesting
 * range of a run is the few percent between its fastest and slowest kilometre,
 * and a zero-based bar squashes that into invisibility. The floor is 25% of
 * the track so the fastest split still reads as a bar, not as nothing.
 */
@Composable
private fun PaceBar(
    paceSeconds: Double?,
    slowestPaceSeconds: Double?,
    fastestPaceSeconds: Double?,
    fasterThanAverage: Boolean,
    modifier: Modifier = Modifier,
) {
    if (paceSeconds == null || slowestPaceSeconds == null || fastestPaceSeconds == null) {
        Spacer(modifier = modifier.height(6.dp))
        return
    }

    val minFraction = 0.25
    val span = slowestPaceSeconds - fastestPaceSeconds
    val fraction = if (span <= 0) {
        1.0
    } else {
        minFraction + (1 - minFraction) * ((paceSeconds - fastestPaceSeconds) / span)
    }
    val color = if (fasterThanAverage) {
        WorkoutColor
    } else {
        MaterialTheme.colorScheme.primary.copy(alpha = 0.55f)
    }

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(6.dp),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction.coerceIn(minFraction, 1.0).toFloat())
                .fillMaxHeight()
                .clip(RoundedCornerShape(3.dp))
                .background(color),
        )
    }
}
